### PR TITLE
js: FetchSource takes optional custom headers

### PR DESCRIPTION
**Pull request by bdon**
Allow passing of custom headers in FetchSource.

### CODE REVIEW

**Comment by m-mohr** [comment]
This would be very helpful to pass authentication details.

**Comment by bdon** [comment]
@m-mohr I should probably change the public API of `ol-pmtiles` to take an instance of `PMTiles` instead of a raw URL, letting you provide a custom `FetchSource` with headers and/or query params.

File: js/index.ts [hunk 1]
```diff
@@ -292,9 +294,17 @@ export class FetchSource implements Source {
...
+		requestHeaders.set(
+			"Range",
+			"bytes=" + offset + "-" + (offset + length - 1)
+		);
+
 		let resp = await fetch(this.url, {
 			signal: signal,
-			headers: { Range: "bytes=" + offset + "-" + (offset + length - 1) },
+			headers: requestHeaders,
 		});
```
Comment (m-mohr) [inline js/index.ts:308]: Would it makse sense to do something like this?

```js
		let options = {
			signal: signal,
			headers: { Range: "bytes=" + offset + "-" + (offset + length - 1) },
			headers: requestHeaders,
		};
		if (typeof this.prepareOptions === 'function') {
			options = this.prepareOptions(options, this.url);
		}
		let resp = await fetch(this.url, options);
```

this.prepareOptions is a callback that can be passed from the user and I assume the URL is passed by the user with potential query parameters anyway.

Then you have full flexibility.

Alternatively, you leave the full request to users. Then they can even choose the request library, e.g. axios.

Comment (bdon) [reply js/index.ts:308]: custom FetchSource feels more flexible, I would want the default FetchSource to be quite narrow in its capabilities, maybe just adding header or query param for well-known use cases like auth,

The big feature missing right now in FetchSource and what will demand a reworking of the `Source` API is support for `If-Match` ETag headers.

Comment (m-mohr) [reply js/index.ts:308]: Can this interface maybe just look as the interface for https://openlayers.org/en/latest/apidoc/module-ol_source_GeoTIFF.html#~GeoTIFF looks like? There you have a headers field in sourceOptions.

Comment (bdon) [reply js/index.ts:308]: @m-mohr I think having `headers` of type `Object.<string, string> | undefined` is good, as long as the headers is static for the lifetime of the Source object. Is this true in your authentication cases, or does it require some dynamic token to be passed that might have been refreshed in a single session?

Comment (m-mohr) [reply js/index.ts:308]: That should be fine, I think.

Comment (bdon) [reply js/index.ts:308]: We can make `ol-pmtiles` conform to the conventions of the other sourceOptions then - I'll plan to work on this during the STAC sprint

Comment (bdon) [reply js/index.ts]: Merging a basic additional param here that takes a `Headers` object. We can make it look more OL-friendly in `ol-pmtiles`

**Pull request merged**